src/lib/formats/rpk.h

```diff
@@ -0,0 +1,149 @@
+// license:LGPL-2.1+
+// copyright-holders:Michael Zapf
+/***************************************************************************
+
+	rpk.h
+
+	RPK format support
+
+***************************************************************************/
+
+#ifndef MAME_FORMATS_RPK_H
+#define MAME_FORMATS_RPK_H
+
+#pragma once
+
+#include "unzip.h"
+#include "hash.h"
+#include "xmlfile.h"
+
+#include <exception>
+#include <optional>
+
+namespace util {
+/***************************************************************************
+	TYPE DEFINITIONS
+***************************************************************************/
+
+class archive_file;
+class rpk_reader;
+class rpk_file;
+
+// ======================> rpk_socket
+
+class rpk_socket
+{
+	friend class rpk_reader;
+public:
+	enum class socket_type
+	{
+		ROM,
+		RAM,
+		PERSISTENT_RAM
+	};
+
+	// ctor/dtor
+	rpk_socket(rpk_file &rpk, std::string &&id, socket_type type, std::string &&filename, std::optional<hash_collection> &&hashes, std::optional<std::uint32_t> length);
+	rpk_socket(const rpk_socket &) = delete;
+	rpk_socket(rpk_socket &&) = delete;
+	~rpk_socket();
+
+	// accessors
+	const std::string &id() const { return m_id; }
+	socket_type type() const { return m_type; }
+	const std::string &filename() const { return m_filename; }
+	std::uint32_t length() const { return m_length.value(); }
```

Comment:
Do you really want the potential for an exception here?  That seems rather dangerous for a `const` accessor.